Invert page table check in polymorphic hash with no-naked-pointers

samwgoldman opened the pull request:
Hashing will visit the fields of a closure, pushing those fields onto the queue. These fields can be code pointers, which need to be distinguished from block values.

Currently, code pointers are detected by looking them up in the page table. This also serves to identify naked pointers. Both are hashed as opaque pointer values.

In no-naked-pointers mode, there is only the need to detect code pointers. This change conditionally replaces the check based on the no-naked-pointers configuration. In no-naked-pointers mode, we use the Is_in_code_area check from address_class.h instead.

The inverse check still uses the page table, but will not pass for values are are stored outside of the heap. This makes it possible to use polymorphic hash with off-heap values.

__Bonus change__

I also noticed that the Forward_tag logic does a page table lookup. In no-naked-pointers mode, it is safe to remove as the pointed-to address will certainly be some value and not a code pointer.

__Backwards compatibility__

The table lookup could be made more granular, by only looking up the non-immediate fields of a closure. However, as the current hashing process will push all fields into the queue, to do otherwise would perturb the order of hashing operations and affect the result.

If backwards compatibility (and "sideways" compatibility between naked-pointer and no-naked-pointer modes) are not important, I am happy to make this change.

I briefly explored the possibility of communicating the extra bit of information into the queue. I believe that words are only guaranteed to be 2-aligned, and that bit is already used to distinguish long vs. block.

But assuming backwards compatibility should be maintained, in the absence of some clever trick, this change will always interrogate the code fragments table, as the page table was always interrogated before.

__Performance__

I benchmarked this on my workload, by inspecting CPU time as well as stack trace samples. Neither showed statistically significant changes. However, this is not a very good benchmark for two reasons. For one, my workload does not make much use of polymorphic hash. `caml_hash` appears in only 0.25% of samples stack traces, for example. Additionally, my workload does not make use of dynamically loaded code, which would affect the performance trade-off.

__Motivation__

I am pursuing this change not for performance, but rather because I would like to store values outside of the heap while still enjoying the polymorphic functions provided by the standard library. #2079 provides this for polymorphic compare (while also providing a nice performance boost), while this closes the loop with polymorphic hash.

samwgoldman committed:
Invert page table check in polymorphic hash with no-naked-pointers

xavierleroy commented on runtime/hash.c:
```diff
@@ -179,6 +180,20 @@ CAMLexport uint32_t caml_hash_mix_string(uint32_t h, value s)
+#define Is_opaque_pointer(v) (caml_code_fragments_table_lookup((void *)(v)))
+#else
+#define Is_opaque_pointer(v) (!Is_in_value_area(v))
+#endif
```
Maybe I'm missing something, but `caml_code_fragments_table_lookup` is probably as costly as the `Is_in_value_area` test, even with only one code fragment, and definitely more costly as the number of code fragments grows (by dynamic linking).  I'd suggest:
* Use `Classify_addr(v) & In_code_area` to test for code pointers.
* Test for code pointers only when scanning the fields of a closure, because they cannot appear anywhere else.

samwgoldman replied:
Thank you for your helpful advice!

You are right, I should just use the page table. I will try again with this approach instead.

I would much prefer to only test for code pointers when scanning the fields of a closure, as you suggested. I considered this originally, but could not come up with a way to do it without perturbing the hash result, breaking backwards compat.

Because we hash in a breadth-first manner, all fields need to be pushed onto the queue, and when we eventually do get to process the closure's fields, we've lost the context that they came from a closure originally.

Do you maybe have another helpful suggestion up your sleeve for that case? :)

xavierleroy replied:
You got me!  I didn't realize breadth-first traversal complicates my suggestion (of scanning closures differently).  We could tag entries in the queue with a "maybe code pointer / definitely not code pointer" bit, but that might be costly.  So, let's forget about it and just query the page table for `In_code_area` status.

xavierleroy reviewed (changes requested):
The O(n) lookup in the table of code fragments is not viable.  See below for alternatives.

samwgoldman commented:
This was a bit more work than I expected. There were a few changes I needed to make to ensure that the page table lookup would find all relevant code pointers:

1. The `Is_in_code_area` macro only compiled in native mode. In bytecode mode, neither `caml_code_area_start` nor `caml_code_area_end` are defined. I just #ifdef'ed my way out of this.
2. meta.c did not update the page table, only the code fragments table.
3. fix_code.c, which adds bytecode loaded from caml_main to the code fragments table, also did not update the page table.

I guess this is what @mshinwell was getting at in #1156 when he said that the tables are fragile and best avoided.

samwgoldman commented:
On the subject of backwards compat, it might not be very important to preserve in the case of hashing closures (or naked pointers), as the addresses are not stable in the first place. Since hashing closures already exhibits this unpredictable behavior, it seems reasonable to push the code pointer check inside a Closure_tag case, which would avoid the page table lookup in every other case. Am I missing something?

samwgoldman committed:
Update Changes

samwgoldman commented:
This PR seems to be "stuck" in the "Changed requested" state, despite my having address the feedback.

samwgoldman commented:
@xavierleroy ping — curious to hear your thoughts on (a) this latest update, which addresses your feedback and (b) whether you agree it's reasonable to interrogate the page table on closures only, as closures already cause the hash value to be unstable.
